webserver/static/js/haarcascade.ts

```diff
@@ -109,181 +117,124 @@ export function buildSAT(
+  for (const scale of scales) {
+    workerPromises.push(
+      new Promise(function (resolve, reject) {
+        let terminationTimer = -1;
+        // Kill the worker if it takes longer than a frame
+        let timeout = 1000 / 8.7;
+        if (WorkerPool.length === 0) {
+          const worker = new Worker("../js/eval-haar-worker.js", {
+            type: "module",
+          });
+          // Copying the HaarCascade data to the worker each frame has significant overhead,
+          // so it's better to just do it once when we init the worker, as the data is constant.
+          worker.postMessage({ type: "init", cascade });
+          WorkerPool.push(worker);
+          // Allow a longer timeout before terminating if the worker has just been created.
+          timeout = 1000;
+        }
+        let worker = WorkerPool.pop() as Worker;
+        let s: number;
+        worker.onmessage = (r) => {
+          clearTimeout(terminationTimer);
+          // Mark resolved and return worker to pool.
+          // console.log(`resolved worker after ${new Date().getTime() - s}, return to pool`);
+          WorkerPool.push(worker);
+          resolve(r.data as ROIFeature[]);
+        };
+        s = new Date().getTime();
+
+        // NOTE(jon): The work can be reasonably evenly divided by each scale we want to search for
+        //  features at.  It's possible that we could make some of the scales happen in serial in a
+        //  single worker, and have less workers, as some scales always take much longer than others.
+        worker.postMessage({
+          type: "eval",
           scale,
           frameWidth,
-          frameHeight
-        );
-        if (ev > 999) {
-          let r = new ROIFeature();
-          r.flavor = "Face";
-          r.x0 = x - scale;
-          r.y0 = y - scale;
-          r.x1 = x + scale;
-          r.y1 = y + scale;
-          let didMerge = false;
-
-          for (let k = 0; k < result.length; k++) {
-            if (result[k].tryMerge(r.x0, r.y0, r.x1, r.y1)) {
-              didMerge = true;
-              break;
-            }
-          }
+          frameHeight,
+          satData,
+          s,
+        });
+        // Terminate the worker if it takes too long?
+        terminationTimer = setTimeout(() => {
+          console.log(
+            `terminating slow worker after ${new Date().getTime() - s}`
+          );
+          worker.terminate();
+          reject([]);
+        }, timeout);
+      })
+    );
+  }
+  try {
+    let results: ROIFeature[][] = await Promise.all(
+      workerPromises as Promise<ROIFeature[]>[]
+    );
+    const allResults = results.reduce(
+      (acc: ROIFeature[], curr: ROIFeature[]) => {
+        acc.push(...curr);
+        return acc;
+      },
+      []
+    );
 
-          if (!didMerge) {
-            result.push(r);
-          }
+    // Merge all boxes.  I *think* this has the same result as doing this work in serial.
+    for (const r of allResults) {
```

Comment:
i think you should do this in reverse, i.e. biggest scale faces first. otherwise you can get a face in a face